### PR TITLE
[RFC] New Decorations API

### PR DESCRIPTION
**Pull request by vigoux**
As discussed with @bfredl , here is an updated version of the new Decorations API.

This removes some functions :

- `nvim_buf_get_virtual_text` : replace by `nvim_buf_get_extmark*`
- `nvim__buf_add_decoration` : replaced by `nvim_buf_set_extmark`

Also update treesitter integration to use this new interface.

### CODE REVIEW

File: src/nvim/api/buffer.c [hunk 1]
```diff
@@ -1309,18 +1355,115 @@ Integer nvim_buf_set_extmark(Buffer buffer, Integer ns_id, Integer id,
...
+        case kObjectTypeString:
+          hl_group = v->data.string;
+          hl_id = syn_check_group((char_u *)(hl_group.data), (int)hl_group.size);
+          break;
+        case kObjectTypeInteger:
+          hl_id = (int)v->data.integer;
+          if (STRLEN(syn_id2name(hl_id)) == 0) {
+            api_set_error(err, kErrorTypeValidation, "hl_group is not a valid id.");
+            goto error;
+          }
```
Comment (vigoux) [inline src/nvim/api/buffer.c:1411]: @bfredl should I have this test or just crudely use the id ?

**Comment by bfredl** [comment]
I think we should attempt to get this merged quite soon. We can modify the API later if needed, and add more tests as we find eventual bugs :]

**Comment by vigoux** [comment]
Appveyor failure does not seem to be related to this PR though

**Comment by tjdevries** [comment]
Did the docs get regenerated for this PR? Should they have been?

File: src/nvim/api/buffer.c [hunk 2]
```diff
@@ -1181,10 +1229,13 @@ ArrayOf(Integer) nvim_buf_get_extmark_by_id(Buffer buffer, Integer ns_id,
...
 /// @param opts  Optional parameters. Keys:
 ///          - limit:  Maximum number of marks to return
+/// @param details Wether to include the details dict
 /// @param[out] err   Error details, if any
 /// @return List of [extmark_id, row, col] tuples in "traversal order".
-Array nvim_buf_get_extmarks(Buffer buffer, Integer ns_id, Object start,
-                            Object end, Dictionary opts, Error *err)
+Array nvim_buf_get_extmarks(Buffer buffer, Integer ns_id,
+                            Object start, Object end,
+                            Dictionary opts, Boolean details,
```
Comment (tjdevries) [inline src/nvim/api/buffer.c:1237]: Can `details` be included in the `opts`? I don't think I understand why they are in different places.

**Pull request merged**